pkg/csi_driver/multishare_stateful_controller.go

```diff
@@ -0,0 +1,297 @@
+}
+
+func (m *MultishareStatefulController) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	_, project, location, instanceName, shareName, err := parseMultishareVolId(req.VolumeId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	klog.V(4).Infof("DeleteVolume called for multishare with request %+v", req)
+
+	siName := util.ShareToShareInfoName(shareName)
+	shareInfo, err := m.shareLister.Get(siName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, status.Errorf(codes.Internal, "error deleting volume %q due to informer error: %s", req.VolumeId, err.Error())
+		}
+		// check with api if share exist
+		klog.V(6).Infof("shareInfo %s DNE in cache, checking if share is already deleted", siName)
+		_, err := m.cloud.File.GetShare(ctx, &file.Share{
+			Parent: &file.MultishareInstance{
+				Project:  project,
+				Location: location,
+				Name:     instanceName,
+			},
+			Name: shareName,
+		})
+		if err != nil {
+			// If share not found, proceed to instance/shrink check.
+			if file.IsNotFoundErr(err) {
+				return &csi.DeleteVolumeResponse{}, nil
+			}
+
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return nil, status.Errorf(codes.Aborted, "waiting to express intent for volume %s to be deleted", req.VolumeId)
+	}
+
+	if shareInfo.DeletionTimestamp == nil {
+		if len(shareInfo.Finalizers) == 0 {
+			klog.Errorf("shareInfo %s shouldn't have no finalizer before deletion marking", siName)
+			return nil, status.Errorf(codes.Internal, "error deleting volume %s due to driver state error", req.VolumeId)
+		}
+		// delete shareInfo
+		err := m.deleteShareInfo(ctx, siName)
+		if err != nil {
+			klog.Errorf("error marking the shareInfo object as deleted: %s", err.Error())
+		}
+		return nil, status.Errorf(codes.Aborted, "expressed intent for volume %s to be deleted, waiting.", req.VolumeId)
+	}
+
+	if shareInfo.Status == nil {
+		klog.Errorf("shareInfo %s marked to be deleted but shareInfo.Status == nil", siName)
+		return nil, status.Errorf(codes.Aborted, "waiting for volume %s to be deleteed.", req.VolumeId)
+	}
+
+	if shareInfo.Status.ShareStatus == v1alpha1.DELETED {
+		// remove finalizer and return success
+		klog.V(6).Infof("trying to remove finalizer from %s because share deleted", siName)
+		shareInfoClone := shareInfo.DeepCopy()
+		shareInfoClone.Finalizers = []string{}
+		shareInfo, err = m.updateShareInfo(ctx, shareInfoClone)
+		if err != nil {
+			klog.Errorf("failed to remove finalizer from %s: %s", siName, err.Error())
+			return nil, status.Errorf(codes.Internal, "error deleting volume %s due to failed internal update", req.VolumeId)
+		}
+		return &csi.DeleteVolumeResponse{}, nil
+	}
+
+	if shareInfo.Status.Error != "" {
+		return nil, status.Errorf(codes.Internal, "internal error: %s", shareInfo.Status.Error)
+	}
+
+	return nil, status.Errorf(codes.Aborted, "waiting for share %s to be deleted", req.VolumeId)
```

Comment:
be clear about the resource name.

if we are using share then we should use shareInfo.spec.sharename